include/deal.II/matrix_free/assemble_operator.h

```diff
@@ -0,0 +1,278 @@
+
+#pragma once
+
+#include <deal.II/lac/vector.h>
+
+#include <boost/graph/adjacency_list.hpp>
+#include <boost/graph/graph_selectors.hpp>
+#include <boost/graph/minimum_degree_ordering.hpp>
+#include <boost/graph/sequential_vertex_coloring.hpp>
+#include <boost/graph/smallest_last_ordering.hpp>
+#include <boost/property_map/property_map.hpp>
+#include <boost/property_map/shared_array_property_map.hpp>
+
+#include <set>
+#include <vector>
+
+// TODO distributed graph
+
+namespace dealii
+{
+  namespace MatrixFreeUtilities
+  {
+    // Typedefs for boost graph types.
+    namespace
+    {
+      using Graph =
+        boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS>;
+      using vertex_descriptor  = boost::graph_traits<Graph>::vertex_descriptor;
+      using vertices_size_type = boost::graph_traits<Graph>::vertices_size_type;
+      using vertex_index_map =
+        boost::property_map<Graph, boost::vertex_index_t>::const_type;
+      using Edge = std::pair<vertices_size_type, vertices_size_type>;
+    } // namespace
+
+    /**
+     * Cache the graph generated from a sparsity pattern for later computations.
+     * This is useful if the matrix does not change its pattern.
+     * // TODO cache entries
+     */
+    struct GraphCache
+    {
+      Graph                           graph;
+      std::vector<vertices_size_type> color_vec;
+      boost::iterator_property_map<vertices_size_type *, vertex_index_map>
+                                                   coloring;
+      vertices_size_type                           num_colors;
+      std::vector<std::vector<vertices_size_type>> entries_per_color;
+      unsigned int                                 max_entries_per_row = 0;
+
+      /**
+       * Create a boost graph (V,E) from the provided sparsity pattern.
+       * The vertices V are given by the column indices.
+       * An edge (v1,v2) is an element of E, if there is a row r,
+       * such that (r,v1) and (r,v2) are both entries of the sparsity pattern.
+       *
+       * @param <tt>pattern</tt> the sparsity pattern for which the graph should be built.
+       * Pattern needs to provide: n_rows(), n_cols(), begin(r), end(r),
+       * iterator->row(), iterator->colum().
+       */
+      template <typename Pattern>
+      void
+      make_graph(const Pattern &pattern)
+      {
+        std::set<Edge> edges;
+        for (unsigned int r = 0; r < pattern.n_rows(); r++)
+          {
+            for (auto it = pattern.begin(r); it != pattern.end(r); it++)
+              {
+                for (auto it2 = it; it2 != pattern.end(r); it2++)
+                  {
+                    AssertDimension(r, it->row());
+                    AssertDimension(r, it2->row());
+                    auto c1 = it->column();
+                    auto c2 = it2->column();
+                    if (c1 > c2)
+                      std::swap(c1, c2);
+                    edges.insert(std::make_pair(c1, c2));
```

Comment:
I think this line limits you to symmetric-sized matrices. I would not swap and rather make the index sets disjunct:
```suggestion
                    edges.insert(std::make_pair(c1, c2+pattern.n_rows())); // or similar
```

Reply:
Both c1 and c2 refer to column indices, so swapping is fine. It's only to avoid duplicate entries in the undirected graph.